### PR TITLE
Refactor chat presence response

### PR DESCRIPTION
**Pull request by notbakaneko**
Now just uses `ChannelTransformer`.
Some additional moderator checks should be fixed now as well.

preloaded users stored in `request()->attributes` for now; this should be moved to a container object if this kind of lookup is useful for other stuff.

### CODE REVIEW

File: app/Models/Chat/Channel.php [hunk 1]
```diff
@@ -387,6 +426,16 @@ public function hasUser(User $user)
         ])->exists();
     }
 
+    public function setUserChannelFor(User $user, UserChannel $userChannel)
+    {
+        if ($userChannel->user_id !== $user->getKey()) {
+            throw new InvariantException('userChannel does not belong to the same user.');
+        }
+
+        $userChannel->setRelation('user', $user);
```
Comment (nanaya) [inline app/Models/Chat/Channel.php:435]: this can be done as part of `userChannelFor`?

File: app/Models/Chat/Channel.php [hunk 2]
```diff
@@ -387,6 +426,16 @@ public function hasUser(User $user)
...
+        $this->userChannel = $userChannel;
```
Comment (nanaya) [inline app/Models/Chat/Channel.php:436]: store in an array keyed by user id? Can then also be set by the getter function

File: app/Models/Chat/Channel.php [hunk 3]
```diff
@@ -387,6 +426,16 @@ public function hasUser(User $user)
         ])->exists();
     }
 
+    public function setUserChannelFor(User $user, UserChannel $userChannel)
+    {
+        if ($userChannel->user_id !== $user->getKey()) {
```
Comment (nanaya) [inline app/Models/Chat/Channel.php:431]: probably not going to make much sense but can use check of `channel_id`

File: app/Http/Controllers/ChatController.php [hunk 4]
```diff
@@ -26,19 +26,17 @@ public function index()
...
-            $canMessage = priv_check('ChatStart', $targetUser)->can();
-
             $channel = Channel::findPM($targetUser, $user);
-            optional($channel)->addUser($user);
+            $channel?->addUser($user);
 
             $json['send_to'] = [
-                'can_message' => $canMessage,
-                'channel_id' => optional($channel)->getKey(),
+                'can_message' => $channel?->canMessage($user) ?? false,
```
Comment (nanaya) [inline app/Http/Controllers/ChatController.php:38]: haven't checked but this seem to be always false for new channels? (`?sendto=user` where there's no pm channel with the `user` yet)

**Review by nanaya — CHANGES_REQUESTED** [review]

File: app/Transformers/Chat/ChannelTransformer.php [hunk 5]
```diff
@@ -50,6 +52,11 @@ public function includeLastMessageId(Channel $channel)
         return $this->primitive($channel->last_message_id);
     }
 
+    public function includeLastReadId(Channel $channel)
+    {
+        return $this->primitive($channel->lastReadIdFor($this->user));
+    }
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply app/Transformers/Chat/ChannelTransformer.php:58]: I was wonder whether it'd be better to change the whole response format to `UserChannel[]` instead 👀

[the remark this answers is not in the document]
Comment (nanaya) [reply app/Transformers/Chat/ChannelTransformer.php:58]: An alternative is `current_user_attributes` like currently in BeatmapDiscussion and BeatmapsetCompact :thinking:

**Review by nanaya — CHANGES_REQUESTED** [review]

**Review by nanaya — APPROVED** [review]

**Pull request merged**